File: api/js/src/validator/normalization.ts
import { layoutButtons } from "../layout.js";
import type { InlineKeyboardButton } from "../types/buttons.js";
import type { KeyboardInput } from "./types.js";

export interface NormalizedButtonRef {
  button: InlineKeyboardButton;
  rowIndex: number;
  columnIndex: number;
  flatIndex: number;
}

export interface NormalizedKeyboard {
  rows: InlineKeyboardButton[][];
  flat: NormalizedButtonRef[];
  rawButtons: InlineKeyboardButton[];
  buttonsPerRow: number;
  autoWrapMaxChars: number;
  isEmpty: boolean;
}

/**
 * Normalizes keyboard state once for all rules (layout + indexed refs).
 */
export function normalizeKeyboard(input: KeyboardInput): NormalizedKeyboard {
  const { buttons, buttonsPerRow, autoWrapMaxChars } = input;
  const rows = layoutButtons(buttons, buttonsPerRow, autoWrapMaxChars);
  const flat: NormalizedButtonRef[] = [];
  let flatIndex = 0;

  for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
    const row = rows[rowIndex];
    if (!row) continue;
    for (let columnIndex = 0; columnIndex < row.length; columnIndex++) {
      const button = row[columnIndex];
      if (!button) continue;
      flat.push({ button, rowIndex, columnIndex, flatIndex });
      flatIndex += 1;
    }
  }

  return {
    rows,
    flat,
    rawButtons: buttons,
    buttonsPerRow,
    autoWrapMaxChars,
    isEmpty: flat.length === 0,
  };
}

File: api/python/telegram_inline_keyboard_builder/validator/normalization.py
from __future__ import annotations

from ..core.layout import layout_buttons
from .types import KeyboardInput, NormalizedButtonRef, NormalizedKeyboard


def normalize_keyboard(input_data: KeyboardInput) -> NormalizedKeyboard:
    """Normalize keyboard state once for all rules (layout + indexed refs)."""
    buttons = input_data["buttons"]
    buttons_per_row = input_data["buttons_per_row"]
    auto_wrap_max_chars = input_data["auto_wrap_max_chars"]
    rows = layout_buttons(buttons, buttons_per_row, auto_wrap_max_chars)
    flat: list[NormalizedButtonRef] = []
    flat_index = 0

    for row_index, row in enumerate(rows):
        for column_index, button in enumerate(row):
            flat.append(
                NormalizedButtonRef(
                    button=button,
                    row_index=row_index,
                    column_index=column_index,
                    flat_index=flat_index,
                )
            )
            flat_index += 1

    return NormalizedKeyboard(
        rows=rows,
        flat=flat,
        raw_buttons=buttons,
        buttons_per_row=buttons_per_row,
        auto_wrap_max_chars=auto_wrap_max_chars,
        is_empty=len(flat) == 0,
    )
